File: flutterapp/lib/recipe_sheet.dart
import 'package:flutter/material.dart';

import 'theme.dart';

class RecipeSheet extends StatelessWidget {
  final bool collapsed;
  final VoidCallback? onTap;

  RecipeSheet({required this.collapsed, this.onTap});

  @override
  Widget build(BuildContext context) {
    final tabs = [
      for (var label in ['INGREDIENTS ', 'STEPS'])
        Tab(
          icon: Text(label, style: Theme.of(context).textTheme.subtitle1),
        ),
    ];

    return DefaultTabController(
      length: 2,
      child: collapsed
          ? Container(
              decoration: BoxDecoration(
                color: basilBackground,
                border:
                    Border(top: BorderSide(color: basilLightGreen, width: 2)),
              ),
              child: TabBar(
                tabs: tabs,
                indicatorColor: Colors.transparent,
                onTap: (_) {
                  onTap?.call();
                },
              ),
            )
          : Container(
              color: basilBackground.withOpacity(0.5),
              // Tried SafeArea here, but that doesn't seem to work!?
              child: Padding(
                padding:
                    EdgeInsets.fromLTRB(0, AppBar().preferredSize.height, 0, 0),
                child: Column(
                  children: [
                    TabBar(
                      tabs: tabs,
                      indicatorColor: basilGreen,
                    ),
                    Expanded(
                      child: TabBarView(
                        children: [RecipeIngredients(), RecipeSteps()],
                      ),
                    )
                  ],
                ),
              ),
            ),
    );
  }
}

class Ingredient {
  String name;
  String amount;
  Ingredient(this.name, this.amount);
}

class RecipeIngredients extends StatelessWidget {
  const RecipeIngredients({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final ingredients = [
      Ingredient('Basil', '6 tbsp'),
      Ingredient('Gluten-free spaghetti', '2 cups'),
      Ingredient('Garlic', '1 tbsp'),
      Ingredient('Ricotta', '4 cups'),
      Ingredient('Kale', '3 cups'),
      Ingredient('Red Pepper Flakes', '1 tbsp'),
      Ingredient('Extra Virgin Olive Oil', '1 tbsp'),
      Ingredient('Salt', '1 tbsp'),
      Ingredient('Pine Nuts', '2 tbsp'),
    ];

    return Container(
      color: Colors.transparent,
      child: Column(children: [
        for (var ingredient in ingredients)
          Padding(
            padding: const EdgeInsets.fromLTRB(8, 16, 8, 16),
            child: Row(
              children: [
                SizedBox(width: 16),
                Icon(
                  Icons.add_circle_outline,
                  color: basilGreen,
                ),
                SizedBox(width: 16),
                Expanded(child: Text(ingredient.name)),
                SizedBox(width: 16),
                Text(ingredient.amount),
                SizedBox(width: 16),
              ],
            ),
          )
      ]),
    );
  }
}

class RecipeSteps extends StatelessWidget {
  const RecipeSteps({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final textTheme = Theme.of(context).textTheme;
    return Padding(
      padding: const EdgeInsets.fromLTRB(16, 16, 64, 0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisAlignment: MainAxisAlignment.start,
        children: <Widget>[
          Text(
            'Chop the Pesto',
            style: textTheme.headline6,
          ),
          Text(
            'Place the basil leaves, garlic and pine nuts on a sturdy cutting board.',
            style: textTheme.headline6!.copyWith(fontWeight: FontWeight.normal),
          ),
        ],
      ),
    );
  }
}

File: composedesktopapp/src/main/kotlin/RecipeSheet.kt
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp

@Composable
fun RecipeSheet(
    modifier: Modifier = Modifier,
) {
    MaterialTheme(typography = MyTypography) {
        Column(
            modifier = modifier.background(basilBackground).fillMaxHeight()
        ) {
            RecipeIngredients()
        }
    }
}

data class Ingredient(val name: String, val amount: String)


@Composable
fun RecipeIngredients(
    modifier: Modifier = Modifier,
) {
    val ingredients = listOf(
        Ingredient("Basil", "6 tbsp"),
        Ingredient("Gluten-free spaghetti", "2 cups"),
        Ingredient("Garlic", "1 tbsp"),
        Ingredient("Ricotta", "4 cups"),
        Ingredient("Kale", "3 cups"),
        Ingredient("Red Pepper Flakes", "1 tbsp"),
        Ingredient("Extra Virgin Olive Oil", "1 tbsp"),
        Ingredient("Salt", "1 tbsp"),
        Ingredient("Pine Nuts", "2 tbsp"),
    )

    Column {
        for (ingredient in ingredients) {
            Row(
                Modifier
                    .padding(start = 8.dp, top = 16.dp, end = 8.dp, bottom = 16.dp)
            ) {
                Spacer(Modifier.width(16.dp))
                // TODO: Add icon
                Text(
                    ingredient.name,
                    modifier = Modifier.weight(1f),
                    style = MaterialTheme.typography.subtitle1
                )
                Spacer(Modifier.width(16.dp))
                Text(
                    ingredient.amount,
                    style = MaterialTheme.typography.subtitle1
                )
                Spacer(Modifier.width(16.dp))
            }
        }
    }
}